File: 278.first-bad-version.py
#
# @lc app=leetcode id=278 lang=python3
#
# [278] First Bad Version
#

# @lc code=start
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
def isBadVersion(version):
    return [False,False,False,False,True,True][version]

class Solution:
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        left = 1
        right = n
        middle = 1
        while left < right:
            middle = (left+right) // 2
            if isBadVersion(middle):
                right = middle
            else:
                left = middle + 1
        return middle
            
        
# @lc code=end

if __name__ == "__main__":
    print(Solution().firstBadVersion(5))

File: 278.first-bad-version.go
/*
 * @lc app=leetcode id=278 lang=golang
 *
 * [278] First Bad Version
 */

// @lc code=start
/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */
// 二分查找的几个关键点
// 1. left = middle+1
// 2. return right

func firstBadVersion(n int) int {
	left := 1
	right := n
	var middle int
	for left < right {
		middle = (left + right) / 2
		if isBadVersion(middle) {
			right = middle
		} else {
			left = middle + 1
		}
	}
	return right
}

// @lc code=end

